services/src/datasets/storage.rs

```diff
@@ -0,0 +1,273 @@
+use geoengine_operators::source::OgrSourceDataset;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use std::path::PathBuf;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSet {
+    pub id: DataSetId,
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo, // TODO: custom type instead of reusing existing one?
+    pub source_operator: String, // TODO: enum?
+}
+
+impl From<AddDataSet> for DataSet {
+    fn from(value: AddDataSet) -> Self {
+        Self {
+            id: DataSetId::Internal(InternalDataSetId::new()),
+            name: value.name,
+            description: value.description,
+            data_type: value.data_type,
+            source_operator: value.source_operator,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AddDataSet {
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo,    // TODO: custom type?
+    pub source_operator: String, // TODO: enum?
+}
+
+impl UserInput for AddDataSet {
+    fn validate(&self) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ImportDataSet {
+    pub name: String,
+    pub description: String,
+    pub data_type: LayerInfo,
+    pub source_operator: String,
+}
+
+impl From<ImportDataSet> for DataSet {
+    fn from(value: ImportDataSet) -> Self {
+        DataSet {
+            id: DataSetId::Internal(InternalDataSetId::new()),
+            name: value.name,
+            description: value.description,
+            data_type: value.data_type,
+            source_operator: value.source_operator,
+        }
+    }
+}
+
+impl UserInput for ImportDataSet {
+    fn validate(&self) -> Result<()> {
+        // TODO
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSetProviderListing {
+    pub id: DataSetProviderId,
+    pub name: String,
+    pub description: String,
+    // more meta data (number of data sets, ...)
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum AddDataSetProvider {
+    AddMockDataSetProvider(AddMockDataSetProvider),
+    // TODO: geo catalog, wcs, ...
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct AddMockDataSetProvider {
+    pub data_sets: Vec<DataSet>,
+}
+
+impl UserInput for AddDataSetProvider {
+    fn validate(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct DataSetProviderListOptions {
+    // TODO: filter
+    pub offset: u32,
+    pub limit: u32,
+}
+
+impl UserInput for DataSetProviderListOptions {
+    fn validate(&self) -> Result<()> {
+        todo!()
+    }
+}
+
+#[allow(clippy::large_enum_variant)] // TODO: box?
+pub enum DataSetLoadingInfo {
+    Raster(RasterLoadingInfo),
+    Vector(VectorLoadingInfo),
+}
+
+pub struct GdalLoadingInfo {
+    pub file: PathBuf,
+}
```

Comment:
Shouldn't this be defined in `operators`?

Reply:
yes this should be in the GdalSource

Reply:
this is more of a placeholder right now. It should hold the information necessary to create the actual GdalDataSetInformationProvider. It should be passed to the datasetdb when adding a data set. It will have functionality outside of the scope of the gdal source (like resolving time intervals from a database). That's why it's in the service crate.